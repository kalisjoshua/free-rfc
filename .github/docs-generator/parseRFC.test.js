const fs = require("fs")
const path = require("path")

const {marked} = require("marked")
const sanitizeHtml = require("sanitize-html")

const {parseRFC} = require("./parseRFC")

const mocks = [1, 2, 3, 4, 5, 6, 7, 8]
  .map((num) => {
    const filepath = path.join("docs-generator", "rfcMocks", `000${num}.md`)
    const md = fs.readFileSync(filepath, "utf8")
    const [title, digits] = md.match(/#\s(.*?)\s+(\d+)[\n$]/).slice(1)
    const id = `#${digits}-${title.toLowerCase().replace(/\s+/g, "-")}`

    return [id, md]
  })

describe("parseRFC", () => {
  it("should throw an error when given a non-string 'id'", () => {
    expect(parseRFC).toThrow("Cannot read properties of undefined (reading 'match')")
  })

  it("should parse a complete and valid RFC markdown", () => {
    expect(parseRFC(...mocks[0])).toMatchSnapshot()
  })

  it("should parse partial RFC markdown", () => {
    expect(parseRFC(...mocks[1])).toMatchSnapshot()
    expect(parseRFC(...mocks[2])).toMatchSnapshot()
    expect(parseRFC(...mocks[3])).toMatchSnapshot()
    expect(parseRFC(...mocks[4])).toMatchSnapshot()
  })

  it("should parse RFC markdown and detect the \"type\"", () => {
    const basic = parseRFC(...mocks[0])
    const replaced = parseRFC(...mocks[5])
    const replacement = parseRFC(...mocks[6])

    expect(basic.type).toBe("root")
    expect(replaced.type).toBe("replaced")
    expect(replacement.type).toBe("replacement")
  })

  it("should skip parsing HTML if the flag is set to true", () => {
    const output = parseRFC(...mocks[0])

    delete output.html

    expect(parseRFC(...mocks[0], true)).toEqual(output)
  })
})
