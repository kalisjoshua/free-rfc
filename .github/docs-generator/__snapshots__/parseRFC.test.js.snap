// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseRFC should parse a complete and valid RFC markdown 1`] = `
Object {
  "html": "<h1><a href=\\"##0001-mock-rfc\\">0001 - Mock RFC 0001</a></h1>
<ul>
<li>Discussion: <a href=\\"https://example.com/discussion\\">https://example.com/discussion</a></li>
<li>Replaces: none</li>
<li>Replaced by: none</li>
<li>Topics: mock, test</li>
</ul>
<h2 class=\\"Summary\\">Summary</h2>
<p>Lorem ipsum dolor sit amet.</p>
",
  "id": "#0001-mock-rfc",
  "replacedBy": "none",
  "replaces": "none",
  "title": "Mock RFC 0001",
  "topics": Array [
    "mock",
    "test",
  ],
  "type": "root",
}
`;

exports[`parseRFC should parse partial RFC markdown 1`] = `
Object {
  "html": "<h1><a href=\\"##0002-mock-rfc\\">0002 - Mock RFC 0002</a></h1>
<ul>
<li>Replaces: none</li>
<li>Replaced by: none</li>
<li>Topics: mock, test</li>
</ul>
<h2 class=\\"Summary\\">Summary</h2>
<p>Lorem ipsum dolor sit amet.</p>
",
  "id": "#0002-mock-rfc",
  "replacedBy": "none",
  "replaces": "none",
  "title": "Mock RFC 0002",
  "topics": Array [
    "mock",
    "test",
  ],
  "type": "root",
}
`;

exports[`parseRFC should parse partial RFC markdown 2`] = `
Object {
  "html": "<h1><a href=\\"##0003-mock-rfc\\">0003 - Mock RFC 0003</a></h1>
<ul>
<li>Replaced by: none</li>
<li>Topics: mock, test</li>
</ul>
<h2 class=\\"Summary\\">Summary</h2>
<p>Lorem ipsum dolor sit amet.</p>
",
  "id": "#0003-mock-rfc",
  "replacedBy": "none",
  "title": "Mock RFC 0003",
  "topics": Array [
    "mock",
    "test",
  ],
  "type": "replacement",
}
`;

exports[`parseRFC should parse partial RFC markdown 3`] = `
Object {
  "html": "<h1><a href=\\"##0004-mock-rfc\\">0004 - Mock RFC 0004</a></h1>
<ul>
<li>Topics: mock, test</li>
</ul>
<h2 class=\\"Summary\\">Summary</h2>
<p>Lorem ipsum dolor sit amet.</p>
",
  "id": "#0004-mock-rfc",
  "title": "Mock RFC 0004",
  "topics": Array [
    "mock",
    "test",
  ],
  "type": "replacement",
}
`;

exports[`parseRFC should parse partial RFC markdown 4`] = `
Object {
  "html": "<h1><a href=\\"##0005-mock-rfc\\">0005 - Mock RFC 0005</a></h1>
<h2 class=\\"Summary\\">Summary</h2>
<p>Lorem ipsum dolor sit amet.</p>
",
  "id": "#0005-mock-rfc",
  "title": "Mock RFC 0005",
  "type": "replacement",
}
`;
